solarforecastarbiter/metrics/evaluator.py

```diff
@@ -0,0 +1,177 @@
+        obs_values = fill_func(observations.value, observations.quality_flag)
+    else:
+        obs_values = observations.values
+
+    # Preprocessing forecasts
+    fx_context = metrics_context['preprocessing']['forecasts']
+    fx_method = fx_context['fill_method']
+    if fx_method in context.supported_fill_functions():
+        fill_func = context._FILL_FUNCTIONS_MAP[fx_method]
+        fx_values = fill_func(forecasts)
+    else:
+        fx_values = forecasts
+
+    # Copy preprocessed timeseries to result
+    if metrics_context['results']['timeseries']['observations']:
+        result['timeseries']['observations'] = obs_values
+
+    if metrics_context['results']['timeseries']['forecasts']:
+        result['timeseries']['forecasts'] = fx_values
+
+    # Calculate metrics
+    for metric, do_calc in metrics_context['metrics'].items():
+        if do_calc and metric in context.supported_metrics():
+
+            # For all
+            metric_func = context._METRICS_MAP[metric]
+            result['metrics']['total'][metric] = metric_func(obs_values,
+                                                             fx_values)
+
+            # Group by calculations
+            group_context = metrics_context['results']['groupings']
+            for groupby, do_group in group_context.items():
+                if do_group and groupby in context.supported_groupings():
+
+                    if groupby not in result['metrics'].keys():
+                        result['metrics'][groupby] = {}
+
+                    result['metrics'][groupby][metric] = \
+                        evaluate_by_group(obs_values,
+                                          fx_values,
+                                          groupby,
+                                          metric_func)
+
+    return result
+
+
+def evaluate_by_group(observation, forecast, groupby, metric_func):
+    """Evaluate the performance according to the groupby.
+
+    Parameters
+    ----------
+    observation : pd.Series
+    forecast : pd.Series
+    groupby : string
+        one of the supported types of groupings
+    metric_func : function
+        a py:func:`solarforecastarbiter.metrics` metrics function
+
+    Returns
+    -------
+    pd.Series :
+
+    Raises
+    ------
+    SfaMetricsConfigError if `groupby` is not a supported groupby type.
+    """
+    result = None
+
+    # Verify timestamps
+    assert np.array_equal(observation.index, forecast.index)
+
+    # Group into single DataFrame
+    df = pd.DataFrame({'observation': observation,
+                       'forecast': forecast},
+                      index=observation.index)
+
+    # Determine proper groupby
+    if groupby == 'month':
+        df_group = df.groupby(df.index.month)
+    elif groupby == 'weekday':
+        df_group = df.groupby(df.index.weekday)
+    elif groupby == 'hour':
+        df_group = df.groupby(df.index.hour)
+    elif groupby == 'date':
+        df_group = df.groupby(df.index.date)
```

Comment:
These groupby operations will need an index that is localized to the Site's timezone to work right. 

@alorenzo175 are all modules responsible for handling tz conversions or should the `api.py` module return data that is localized to `Site.timezone`?

Reply:
I think we should probably leave it up to modules. Most probably won't need localtime so and the ``api.py`` data should be UTC localized already.

Reply:
Ok. I'll add the ability to set the timezone in the MetricsContext using the IANA timezone as is used in the Site class. @wholmgren should the timeseries of the processed data use the provided timezone or the original?

Reply:
I think localized to `site.timezone` is preferable for processed data.

Reply:
Understood. That's how I've done it and now just writing some tests. Any thoughts on providing the Site object to the evaluate function along with the forecasts and observations?

Reply:
I suspected that you'd need to get the full `datamodel.Forecast` and `datamodel.Observation` metadata into the context and/or the evaluate function in order to handle resampling and alignment issues. They each have a `.site` attribute.

Reply:
I was thinking to keep the amount of data provided to the metrics evaluator should be minimal but there was no real reason behind that come to think of it. I can change it so that the metadata models are required.

Reply:
At this point I don't have a preference for requiring the full data model objects or just their attributes as necessary. I just thought you'd eventually need so many of their attributes that it would be easier to take the whole package.

Reply:
Added timezone support from the context and fixed a bug with results. Now back to modifying the structure of MetricsContext and MetricsResult.